include/astshim/FrameDict.h

```diff
@@ -0,0 +1,310 @@
+
+    @param[in] baseFrame  base @ref Frame.
+    @param[in] mapping  mapping connecting baseFrame to currentFrame.
+    @param[in] currentFrame  current @ref Frame.
+    @param[in] options  Comma-separated list of attribute assignments.
+
+    @throws std::invalid_argument if both Frames have the same non-empty domain.
+    */
+    explicit FrameDict(Frame const &baseFrame, Mapping const &mapping, Frame const &currentFrame,
+                       std::string const &options = "")
+            : FrameSet(baseFrame, mapping, currentFrame) {
+        _rebuildDict();
+    }
+
+    /**
+    Construct a FrameDict from a FrameSet
+
+    The FrameSet is deep-copied
+
+    @throws std::invalid_argument if two Frames in the FrameSet have the same non-empty domain.
+    */
+    explicit FrameDict(FrameSet const &frameSet)
+            : FrameDict(reinterpret_cast<AstFrameSet *>(frameSet.copy()->getRawPtr())) {}
+    virtual ~FrameDict() {}
+
+    FrameDict(FrameDict const &) = delete;
+    FrameDict(FrameDict &&) = default;
+    FrameDict &operator=(FrameDict const &) = delete;
+    FrameDict &operator=(FrameDict &&) = default;
+
+    /// Return a deep copy of this object.
+    std::shared_ptr<FrameSet> copy() const { return std::static_pointer_cast<FrameDict>(copyPolymorphic()); }
+
+    void addFrame(int iframe, Mapping const &map, Frame const &frame) override {
+        FrameSet::addFrame(iframe, map, frame);
+        _addFrameToDict(frame, FrameSet::getCurrent());
+    }
+
+    /**
+    Add a Frame connected by a Mapping to an existing initial Frame. A variant of FrameSet::addFrame
+    where the initial frame is specified by domain.
+
+    @throws std::invalid_argument if `frame` has a non-empty domain and this FrameDict already
+    contains a Frame with that domain
+    */
+    void addFrame(std::string const &domain, Mapping const &map, Frame const &frame) {
+        addFrame(getIndex(domain), map, frame);
+    }
+
+    /**
+    Get a list of the domain names for all contained Frames that have non-empty domain names.
+
+    The order of the names is unspecified.
+    */
+    std::vector<std::string> getAllDomains() const {
+        std::vector<std::string> domains;
+        domains.reserve(_domainIndexDict.size());
+        for (auto const &item : _domainIndexDict) {
+            domains.emplace_back(item.first);
+        }
+        return domains;
+    }
+
+    using FrameSet::getFrame;
+
+    /**
+    Get a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    std::shared_ptr<Frame> getFrame(std::string const &domain, bool copy = true) const {
+        return FrameSet::getFrame(getIndex(domain), copy);
+    }
+
+    using FrameSet::getMapping;
+
+    //@{
+    /**
+    Get a mapping from one frame to another, with one or both frames specified by domain
+
+    @throw std::range_error if no frame found with the specified from or to domain
+    */
```

Comment:
The docs for most methods look fine to me. However getMapping can't find the original when I specify it as `@ref getMapping(int, int)` -- this appears to be a Doxygen bug, since all other methods are found.  After some time playing around with it, I gave up and left it. I found if I specified `@ref getMapping(int from=BASE, int to=CURRENT)` it stopped generating warnings, but the links it generated were useless, which I consider worse than having the Doxygen warnings.